### PR TITLE
update: upgrade @openzeppelin contract packages to latest 5.0.0 version (#536)

### LINKED ISSUE #536: [OZ 5.x Contract Support] Upgrade @openzeppelin/contracts npm package to latest 5.x
### Problem

The Hedera network supports smart contracts as part of EVM equivalence.
Occasionally items have shown up that surfaced inequivalence at the language or opcode layer.

OZ provides many useful templates that smart contract devs are used to utilizing.
However, it may not be clear what coverage the Hedera Smart Contract Service 2.0 has for them

### Solution

upgrade @openzeppeline/contracts npm package to latest v.5.0.0 in package.json

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by quiet-node**
**Description**:
- Upgrade @openzeppelin/contracts and @openzeppelin/upgradable-contracts to latest v5.0.0
- Modified smart contracts utilizing OZ contracts to match with latest changes
- Fixed warnings
- Recompiled all contracts
- In order to bypass the `upgrade safe` checks from hardhat-updates's .deployProxy() method, contract `OZUUPSUpgradeableV4` is added. This  `OZUUPSUpgradeableV4` is a clone of the `UUPSUpgradable` contract from OZ @v5.x but with an extra `upgradeTo` function. This contract is only for unblocking purpose and should be deprecated in the future release when we upgrade to `ethers@v6.x`

**Related issue(s)**: 

Fixes #536 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol [hunk 1]
```diff
@@ -138,15 +138,6 @@ contract HederaNonFungibleToken is ERC721, Constants {
         return super.transferFrom(from, to, tokenId);
     }
 
-    function safeTransferFrom(address from, address to, uint256 tokenId) public override {
```
Comment (quiet-node) [inline test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:138]: notice: This is being removed as OZ/contract v5.0.0 only allows override safeTransferFrom(address,address,uint256,bytes), and safeTransaferFrom(address,address,uint256) will cause error `Trying to override non-virtual function.`

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Great work, blocking until we cut a 0.6 release branch

**Comment by stefan-stefanooov** [comment]
Nicely done ... I have several branches that I want to rebase on this one. Waiting for the murgure...

File: contracts/proxy-upgrade/exchange/OZUUPSUpgradeableV4.sol [hunk 2]
```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: MIT
```
Comment (Nana-EC) [inline contracts/proxy-upgrade/exchange/OZUUPSUpgradeableV4.sol:1]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: contracts/solidity/inhetitance/Main.sol [hunk 3]
```diff
@@ -14,8 +14,4 @@ contract Main is Base {
     function returnSuper() public view virtual returns (string memory) {
         return super.classIdentifier();
     }
-
-    function destroyContract(address recipient) public {
```
Comment (Nana-EC) [inline contracts/solidity/inhetitance/Main.sol:18]: Please restore.
We can remove it separately but until Cancun let's maintain the coverage

File: local.env [hunk 4]
```diff
@@ -8,6 +8,6 @@ OPERATOR_KEY_A= 302e020100300506032b65700422042091132178e72057a1d7528025956fe39b
 # ECDSA HEX Encoded Private Key
 HEX_PRIVATE_KEY_A= 0x2e1d968b041d84dd120a5860cee60cd83f9374ef527ca86996317ada3d0d03e7
 HEX_PRIVATE_KEY_B= 0x45a5a7108a18dd5013cf2d5857a28144beadc9c70b3bdbd914e38df4e804b8d8
-RETRY_DELAY=3000 # ms
+RETRY_DELAY=4000 # ms
```
Comment (Nana-EC) [inline local.env:11]: Q:why the increase?

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol [hunk 5]
```diff
@@ -158,8 +152,8 @@ contract HederaNonFungibleToken is ERC721, Constants {
         return uint64(nftCount.minted - nftCount.burned);
     }
 
-    function isApprovedOrOwner(address spender, uint256 tokenId) external view returns (bool) {
-        return _isApprovedOrOwner(spender, tokenId);
+    function isApprovedOrOwner(address owner, address spender, uint256 tokenId) external view returns (bool) {
```
Comment (Nana-EC) [inline test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:155]: Q: why this change was something broken or is the name wrong?
Even if the change is needed should it be an additional app or is it appropriate to modify?

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Mostly good.
2 items

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol (hunk 5 above)
Comment (quiet-node) [reply test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:155]: So the `_isAuthorized()` method on line 156 is from OZ/ERC721.sol. In the upgrade, OZ added a new argument for the `_isAuthorized()` method that's why it has this change. I'm not quite clear on what you're asking for here 
> Even if the change is needed should it be an additional app or is it appropriate to modify?

File: local.env (hunk 4 above)
Comment (quiet-node) [reply local.env:11]: The `should be able to execute cryptoTransfer for hbar transfer only` unit test in the `TokenTransferContract Test Suite` test suite was consistently failing with as the data doesn't have enough time to reflect the changes so I decided to increase it for 1000 more ms.

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol (hunk 5 above)
Comment (Nana-EC) [reply test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:155]: Okay, great, mixed it up for something else. LG

File: local.env (hunk 4 above)
Comment (Nana-EC) [reply local.env:11]: I understand but it seems strange that 1 min (3000 ms * 20) isn't enough.
Now we're making it 80s (4000ms *20) something is up.
What is happening that requires this long of retries?

**Review by Nana-EC — APPROVED** [review]
